src/coreclr/jit/flowgraph.cpp

```diff
@@ -436,6 +436,306 @@ BasicBlock* Compiler::fgCreateGCPoll(GCPollType pollType, BasicBlock* block)
+
+                int                  isInitOffset = 0;
+                CORINFO_CONST_LOOKUP flagAddr     = {};
+                if (!info.compCompHnd->getIsClassInitedFlagAddress(call->gtRetClsHnd, &flagAddr, &isInitOffset))
+                {
+                    JITDUMP("getIsClassInitedFlagAddress returned false - bail out.\n")
+                    continue;
+                }
+
+                CORINFO_CONST_LOOKUP staticBaseAddr = {};
+                if ((retValKind == SHRV_STATIC_BASE_PTR) &&
+                    !info.compCompHnd->getStaticBaseAddress(call->gtRetClsHnd, isGc, &staticBaseAddr))
+                {
+                    JITDUMP("getStaticBaseAddress returned false - bail out.\n")
+                    continue;
+                }
+
+                DebugInfo debugInfo = stmt->GetDebugInfo();
+
+                // Split block right before the call tree
+                BasicBlock* prevBb       = block;
+                GenTree**   callUse      = nullptr;
+                Statement*  newFirstStmt = nullptr;
+                block                    = fgSplitBlockBeforeTree(block, stmt, call, &newFirstStmt, &callUse);
+                assert(prevBb != nullptr && block != nullptr);
+
+                // Block ops inserted by the split need to be morphed here since we are after morph.
+                // We cannot morph stmt yet as we may modify it further below, and the morphing
+                // could invalidate callUse.
+                while ((newFirstStmt != nullptr) && (newFirstStmt != stmt))
+                {
+                    fgMorphStmtBlockOps(block, newFirstStmt);
+                    newFirstStmt = newFirstStmt->GetNextStmt();
+                }
+
+                //
+                // Create new blocks. Essentially, we want to transform this:
+                //
+                //   staticBase = helperCall();
+                //
+                // into:
+                //
+                //   if (!isInitialized)
+                //   {
+                //       helperCall(); // we don't use its return value
+                //   }
+                //   staticBase = fastPath;
+                //
+
+                // The initialization check looks like this for JIT:
+                //
+                // *  JTRUE     void
+                // \--*  EQ        int
+                //    +--*  AND       int
+                //    |  +--*  IND       int
+                //    |  |  \--*  CNS_INT(h) long   0x.... const ptr
+                //    |  \--*  CNS_INT   int    1 (bit mask)
+                //    \--*  CNS_INT   int    1
+                //
+                // For NativeAOT it's:
+                //
+                // *  JTRUE     void
+                // \--*  EQ        int
+                //    +--*  IND       nint
+                //    |  \--*  ADD       long
+                //    |     +--*  CNS_INT(h) long   0x.... const ptr
+                //    |     \--*  CNS_INT   int    -8 (offset)
```

Comment:
Just FYI: it is possible to "fold" this at the AOT compiler level by providing a symbol that's offset from the base by a pointer (i. e. `ISymbolNode.Offset => -TARGET_POINTER_SIZE`).

At the point this optimization has run, CSE has done its thing, so it may even be better CQ-wise, though we still have the "reuse reg val" in LSRA.

Reply:
we do here CSE by hands - base is saved to a local to be returned as the result (for non-gc case). Also, presumably a new symbol will increase binary size on naot size? Anyway, would love to experiment on that separetely so it will be easier to detect codegen size changes